sometimes search result will be empty in azure app service

sometimes search result will be empty in azure app service for some instances, such as if the app have 3 instances, one will not able to search but the other 2 instances works fine. there have error logged:

```
[ERROR] DotNetNuke.Services.Exceptions.Exceptions - System.IO.IOException: An unexpected network error occurred.
 
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.FileStream.ReadCore(Byte[] buffer, Int32 offset, Int32 count)
   at System.IO.FileStream.Read(Byte[] array, Int32 offset, Int32 count)
   at Lucene.Net.Store.SimpleFSDirectory.SimpleFSIndexInput.ReadInternal(Byte[] b, Int32 offset, Int32 len)
   at Lucene.Net.Store.BufferedIndexInput.Refill()
   at Lucene.Net.Store.BufferedIndexInput.ReadByte()
   at Lucene.Net.Store.IndexInput.ReadVInt()
   at Lucene.Net.Index.TermBuffer.Read(IndexInput input, FieldInfos fieldInfos)
   at Lucene.Net.Index.SegmentTermEnum.Next()
   at Lucene.Net.Index.SegmentTermEnum.ScanTo(Term term)
   at Lucene.Net.Index.TermInfosReader.Get(Term term, Boolean useCache)
   at Lucene.Net.Index.TermInfosReader.Terms(Term term)
   at Lucene.Net.Index.DirectoryReader.MultiTermEnum..ctor(IndexReader topReader, IndexReader[] readers, Int32[] starts, Term t)
   at Lucene.Net.Index.DirectoryReader.Terms(Term term)
   at Lucene.Net.Search.PrefixTermEnum..ctor(IndexReader reader, Term prefix)
   at Lucene.Net.Search.PrefixQuery.GetEnum(IndexReader reader)
   at Lucene.Net.Search.MultiTermQuery.ConstantScoreAutoRewrite.Rewrite(IndexReader reader, MultiTermQuery query)
   at Lucene.Net.Search.BooleanQuery.Rewrite(IndexReader reader)
   at Lucene.Net.Search.BooleanQuery.Rewrite(IndexReader reader)
   at Lucene.Net.Search.BooleanQuery.Rewrite(IndexReader reader)
   at Lucene.Net.Search.IndexSearcher.Rewrite(Query original)
   at Lucene.Net.Search.Query.Weight(Searcher searcher)
   at Lucene.Net.Search.Searcher.Search(Query query, Filter filter, Collector results)
   at DotNetNuke.Services.Search.Internals.LuceneControllerImpl.Search(LuceneSearchContext searchContext)
   at DotNetNuke.Services.Search.Controllers.SearchControllerImpl.GetSecurityTrimmedResults(SearchQuery searchQuery, LuceneQuery luceneQuery)
   at DotNetNuke.Services.Search.Controllers.SearchControllerImpl.GetResults(SearchQuery searchQuery)
   at DotNetNuke.Services.Search.Controllers.SearchControllerImpl.SiteSearch(SearchQuery searchQuery)
   at DotNetNuke.Web.InternalServices.SearchServiceController.GetBasicViews(SearchQuery searchQuery, Int32& totalHits)
   at DotNetNuke.Web.InternalServices.SearchServiceController.GetGroupedBasicViews(SearchQuery query, SearchContentSource userSearchSource, Int32 portalId)
   at DotNetNuke.Web.InternalServices.SearchServiceController.Preview(String keywords, String culture, Int32 forceWild, Int32 portal)
```

the only workaround is restart that instance and search will back to work.

--- a/DNN Platform/Library/Services/Search/Internals/Constants.cs
+++ b/DNN Platform/Library/Services/Search/Internals/Constants.cs
@@ -118,6 +118,7 @@ namespace DotNetNuke.Services.Search.Internals
         internal const string SearchLastSuccessIndexName = "Search_LastSuccessfulIndexedOn";
         internal const string SearchOptimizeFlagName = "Search_OptimizeIndex";
         internal const string SearchCustomAnalyzer = "Search_CustomAnalyzer";
+        internal const string SearchRetryTimesKey = "Search_RetryTimes";
 
         // misc.
         internal const string TlsSearchInfo = "TLS_SEARCH_INFO";

--- a/DNN Platform/Library/Services/Search/Internals/LuceneControllerImpl.cs
+++ b/DNN Platform/Library/Services/Search/Internals/LuceneControllerImpl.cs
@@ -257,11 +257,6 @@ namespace DotNetNuke.Services.Search.Internals
 
         public LuceneResults Search(LuceneSearchContext searchContext)
         {
-            return Search(searchContext, 0);
-        }
-
-        private LuceneResults Search(LuceneSearchContext searchContext, int retryTimes)
-        {
             Requires.NotNull("LuceneQuery", searchContext.LuceneQuery);
             Requires.NotNull("LuceneQuery.Query", searchContext.LuceneQuery.Query);
             Requires.PropertyNotEqualTo("LuceneQuery", "PageSize", searchContext.LuceneQuery.PageSize, 0);
@@ -282,59 +277,62 @@ namespace DotNetNuke.Services.Search.Internals
             var maxResults = searchContext.LuceneQuery.PageIndex * searchContext.LuceneQuery.PageSize;
             var minResults = maxResults - searchContext.LuceneQuery.PageSize + 1;
 
-            try
+            for (var i = 0; i < _searchRetryTimes; i++)
             {
-                var searcher = GetSearcher();
-                var searchSecurityTrimmer = new SearchSecurityTrimmer(new SearchSecurityTrimmerContext
+                try
+                {
+                    var searcher = GetSearcher();
+                    var searchSecurityTrimmer = new SearchSecurityTrimmer(new SearchSecurityTrimmerContext
                     {
                         Searcher = searcher,
                         SecurityChecker = searchContext.SecurityCheckerDelegate,
                         LuceneQuery = searchContext.LuceneQuery,
                         SearchQuery = searchContext.SearchQuery
                     });
-                searcher.Search(searchContext.LuceneQuery.Query, null, searchSecurityTrimmer);
-                luceneResults.TotalHits = searchSecurityTrimmer.TotalHits;
+                    searcher.Search(searchContext.LuceneQuery.Query, null, searchSecurityTrimmer);
+                    luceneResults.TotalHits = searchSecurityTrimmer.TotalHits;
 
-                if (Logger.IsDebugEnabled)
-                {
-                    var sb = GetSearcResultExplanation(searchContext.LuceneQuery, searchSecurityTrimmer.ScoreDocs, searcher);
-		            Logger.Trace(sb);
+                    if (Logger.IsDebugEnabled)
+                    {
+                        var sb = GetSearcResultExplanation(searchContext.LuceneQuery, searchSecurityTrimmer.ScoreDocs, searcher);
+                        Logger.Trace(sb);
+                    }
+
+                    //Page doesn't exist
+                    if (luceneResults.TotalHits >= minResults)
+                    {
+                        luceneResults.Results = searchSecurityTrimmer.ScoreDocs.Select(match =>
+                            new LuceneResult
+                            {
+                                Document = searcher.Doc(match.Doc),
+                                Score = match.Score,
+                                DisplayScore = GetDisplayScoreFromMatch(match.ToString()),
+                                TitleSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.TitleTag, searchContext.LuceneQuery.TitleSnippetLength),
+                                BodySnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.BodyTag, searchContext.LuceneQuery.BodySnippetLength),
+                                DescriptionSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.DescriptionTag, searchContext.LuceneQuery.TitleSnippetLength),
+                                TagSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.Tag, searchContext.LuceneQuery.TitleSnippetLength),
+                                AuthorSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.AuthorNameTag, searchContext.LuceneQuery.TitleSnippetLength),
+                                ContentSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.ContentTag, searchContext.LuceneQuery.TitleSnippetLength)
+                            }).ToList();
+                    }
+                    break;
                 }
-
-                //Page doesn't exist
-                if (luceneResults.TotalHits < minResults)
-                    return luceneResults;
-
-                luceneResults.Results  = searchSecurityTrimmer.ScoreDocs.Select(match =>
-                    new LuceneResult
-                        {
-                            Document = searcher.Doc(match.Doc),
-                            Score = match.Score,
-                            DisplayScore = GetDisplayScoreFromMatch(match.ToString()),
-                            TitleSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.TitleTag, searchContext.LuceneQuery.TitleSnippetLength),
-                            BodySnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.BodyTag, searchContext.LuceneQuery.BodySnippetLength),
-                            DescriptionSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.DescriptionTag, searchContext.LuceneQuery.TitleSnippetLength),
-                            TagSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.Tag, searchContext.LuceneQuery.TitleSnippetLength),
-                            AuthorSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.AuthorNameTag, searchContext.LuceneQuery.TitleSnippetLength),
-                            ContentSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.ContentTag, searchContext.LuceneQuery.TitleSnippetLength)
-                        }).ToList();
-                return luceneResults;
-            }
-            catch (Exception ex)
-            {
-                if (retryTimes >= _searchRetryTimes)
+                catch (IOException ex)
                 {
-                    throw;
+                    DisposeReaders();
+
+                    if (i == _searchRetryTimes - 1)
+                    {
+                        throw;
+                    }
+
+                    Logger.Error(ex);
+                    Logger.Error($"Search Index Folder Is Not Available: {ex.Message}, Retry {i + 1} time(s).");
+                    Thread.Sleep(100);
                 }
-
-                DisposeReaders();
-                retryTimes++;
-                Logger.Error(ex);
-                Logger.Error($"Search Index Folder Is Not Available: {ex.Message}, Retry {retryTimes} time(s).");
-                Thread.Sleep(100);
-
-                return Search(searchContext, retryTimes);
             }
+
+            return luceneResults;
         }
 
         private string GetHighlightedText(FastVectorHighlighter highlighter, FieldQuery fieldQuery, IndexSearcher searcher, ScoreDoc match, string tag, int length)

--- a/DNN Platform/Library/Services/Search/Internals/LuceneControllerImpl.cs
+++ b/DNN Platform/Library/Services/Search/Internals/LuceneControllerImpl.cs
@@ -299,22 +299,22 @@ namespace DotNetNuke.Services.Search.Internals
                     }
 
                     //Page doesn't exist
-                    if (luceneResults.TotalHits >= minResults)
-                    {
-                        luceneResults.Results = searchSecurityTrimmer.ScoreDocs.Select(match =>
-                            new LuceneResult
-                            {
-                                Document = searcher.Doc(match.Doc),
-                                Score = match.Score,
-                                DisplayScore = GetDisplayScoreFromMatch(match.ToString()),
-                                TitleSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.TitleTag, searchContext.LuceneQuery.TitleSnippetLength),
-                                BodySnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.BodyTag, searchContext.LuceneQuery.BodySnippetLength),
-                                DescriptionSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.DescriptionTag, searchContext.LuceneQuery.TitleSnippetLength),
-                                TagSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.Tag, searchContext.LuceneQuery.TitleSnippetLength),
-                                AuthorSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.AuthorNameTag, searchContext.LuceneQuery.TitleSnippetLength),
-                                ContentSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.ContentTag, searchContext.LuceneQuery.TitleSnippetLength)
-                            }).ToList();
-                    }
+                    if (luceneResults.TotalHits < minResults)
+                        break;
+
+                    luceneResults.Results = searchSecurityTrimmer.ScoreDocs.Select(match =>
+                        new LuceneResult
+                        {
+                            Document = searcher.Doc(match.Doc),
+                            Score = match.Score,
+                            DisplayScore = GetDisplayScoreFromMatch(match.ToString()),
+                            TitleSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.TitleTag, searchContext.LuceneQuery.TitleSnippetLength),
+                            BodySnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.BodyTag, searchContext.LuceneQuery.BodySnippetLength),
+                            DescriptionSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.DescriptionTag, searchContext.LuceneQuery.TitleSnippetLength),
+                            TagSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.Tag, searchContext.LuceneQuery.TitleSnippetLength),
+                            AuthorSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.AuthorNameTag, searchContext.LuceneQuery.TitleSnippetLength),
+                            ContentSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.ContentTag, searchContext.LuceneQuery.TitleSnippetLength)
+                        }).ToList();
                     break;
                 }
                 catch (IOException ex)

--- a/DNN Platform/Library/Services/Search/Internals/LuceneControllerImpl.cs
+++ b/DNN Platform/Library/Services/Search/Internals/LuceneControllerImpl.cs
@@ -58,6 +58,7 @@ namespace DotNetNuke.Services.Search.Internals
         private const string DefaultSearchFolder = @"App_Data\Search";
         private const string WriteLockFile = "write.lock";
         internal const int DefaultRereadTimeSpan = 30; // in seconds
+        private const int DefaultSearchRetryTimes = 5;
         private const int DISPOSED = 1;
         private const int UNDISPOSED = 0;
         #endregion
@@ -74,6 +75,7 @@ namespace DotNetNuke.Services.Search.Internals
         private FastVectorHighlighter _fastHighlighter;
         private readonly object _writerLock = new object();
         private readonly double _readerTimeSpan; // in seconds
+        private readonly int _searchRetryTimes; //search retry times if exception thrown during search process
         private readonly List<CachedReader> _oldReaders = new List<CachedReader>();
         private int _isDisposed = UNDISPOSED;
 
@@ -92,6 +94,7 @@ namespace DotNetNuke.Services.Search.Internals
             if (string.IsNullOrEmpty(folder)) folder = DefaultSearchFolder;
             IndexFolder = Path.Combine(Globals.ApplicationMapPath, folder);
             _readerTimeSpan = hostController.GetDouble(Constants.SearchReaderRefreshTimeKey, DefaultRereadTimeSpan);
+            _searchRetryTimes = hostController.GetInteger(Constants.SearchRetryTimesKey, DefaultSearchRetryTimes);
         }
 
         private void CheckDisposed()
@@ -254,6 +257,11 @@ namespace DotNetNuke.Services.Search.Internals
 
         public LuceneResults Search(LuceneSearchContext searchContext)
         {
+            return Search(searchContext, 0);
+        }
+
+        private LuceneResults Search(LuceneSearchContext searchContext, int retryTimes)
+        {
             Requires.NotNull("LuceneQuery", searchContext.LuceneQuery);
             Requires.NotNull("LuceneQuery.Query", searchContext.LuceneQuery.Query);
             Requires.PropertyNotEqualTo("LuceneQuery", "PageSize", searchContext.LuceneQuery.PageSize, 0);
@@ -274,41 +282,59 @@ namespace DotNetNuke.Services.Search.Internals
             var maxResults = searchContext.LuceneQuery.PageIndex * searchContext.LuceneQuery.PageSize;
             var minResults = maxResults - searchContext.LuceneQuery.PageSize + 1;
 
-            var searcher = GetSearcher();
-            var searchSecurityTrimmer = new SearchSecurityTrimmer(new SearchSecurityTrimmerContext
-                {
-                    Searcher = searcher,
-                    SecurityChecker = searchContext.SecurityCheckerDelegate,
-                    LuceneQuery = searchContext.LuceneQuery,
-                    SearchQuery = searchContext.SearchQuery
-                });
-            searcher.Search(searchContext.LuceneQuery.Query, null, searchSecurityTrimmer);
-            luceneResults.TotalHits = searchSecurityTrimmer.TotalHits;
-
-            if (Logger.IsDebugEnabled)
+            try
             {
-                var sb = GetSearcResultExplanation(searchContext.LuceneQuery, searchSecurityTrimmer.ScoreDocs, searcher);
-		        Logger.Trace(sb);
-            }
-
-            //Page doesn't exist
-            if (luceneResults.TotalHits < minResults)
-                return luceneResults;
-
-            luceneResults.Results  = searchSecurityTrimmer.ScoreDocs.Select(match =>
-                new LuceneResult
+                var searcher = GetSearcher();
+                var searchSecurityTrimmer = new SearchSecurityTrimmer(new SearchSecurityTrimmerContext
                     {
-                        Document = searcher.Doc(match.Doc),
-                        Score = match.Score,
-                        DisplayScore = GetDisplayScoreFromMatch(match.ToString()),
-                        TitleSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.TitleTag, searchContext.LuceneQuery.TitleSnippetLength),
-                        BodySnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.BodyTag, searchContext.LuceneQuery.BodySnippetLength),
-                        DescriptionSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.DescriptionTag, searchContext.LuceneQuery.TitleSnippetLength),
-                        TagSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.Tag, searchContext.LuceneQuery.TitleSnippetLength),
-                        AuthorSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.AuthorNameTag, searchContext.LuceneQuery.TitleSnippetLength),
-                        ContentSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.ContentTag, searchContext.LuceneQuery.TitleSnippetLength)
-                    }).ToList();
-            return luceneResults;
+                        Searcher = searcher,
+                        SecurityChecker = searchContext.SecurityCheckerDelegate,
+                        LuceneQuery = searchContext.LuceneQuery,
+                        SearchQuery = searchContext.SearchQuery
+                    });
+                searcher.Search(searchContext.LuceneQuery.Query, null, searchSecurityTrimmer);
+                luceneResults.TotalHits = searchSecurityTrimmer.TotalHits;
+
+                if (Logger.IsDebugEnabled)
+                {
+                    var sb = GetSearcResultExplanation(searchContext.LuceneQuery, searchSecurityTrimmer.ScoreDocs, searcher);
+		            Logger.Trace(sb);
+                }
+
+                //Page doesn't exist
+                if (luceneResults.TotalHits < minResults)
+                    return luceneResults;
+
+                luceneResults.Results  = searchSecurityTrimmer.ScoreDocs.Select(match =>
+                    new LuceneResult
+                        {
+                            Document = searcher.Doc(match.Doc),
+                            Score = match.Score,
+                            DisplayScore = GetDisplayScoreFromMatch(match.ToString()),
+                            TitleSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.TitleTag, searchContext.LuceneQuery.TitleSnippetLength),
+                            BodySnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.BodyTag, searchContext.LuceneQuery.BodySnippetLength),
+                            DescriptionSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.DescriptionTag, searchContext.LuceneQuery.TitleSnippetLength),
+                            TagSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.Tag, searchContext.LuceneQuery.TitleSnippetLength),
+                            AuthorSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.AuthorNameTag, searchContext.LuceneQuery.TitleSnippetLength),
+                            ContentSnippet = GetHighlightedText(highlighter, fieldQuery, searcher, match, Constants.ContentTag, searchContext.LuceneQuery.TitleSnippetLength)
+                        }).ToList();
+                return luceneResults;
+            }
+            catch (Exception ex)
+            {
+                if (retryTimes >= _searchRetryTimes)
+                {
+                    throw;
+                }
+
+                DisposeReaders();
+                retryTimes++;
+                Logger.Error(ex);
+                Logger.Error($"Search Index Folder Is Not Available: {ex.Message}, Retry {retryTimes} time(s).");
+                Thread.Sleep(100);
+
+                return Search(searchContext, retryTimes);
+            }
         }
 
         private string GetHighlightedText(FastVectorHighlighter highlighter, FieldQuery fieldQuery, IndexSearcher searcher, ScoreDoc match, string tag, int length)

